train_tacotron.py

```diff
@@ -5,21 +5,64 @@
 import torch
 from torch import optim
 from torch.utils.data.dataloader import DataLoader
+from typing import Tuple
 
 from models.tacotron import Tacotron
 from trainer.taco_trainer import TacoTrainer
 from utils import hparams as hp
 from utils.checkpoints import restore_checkpoint
-from utils.dataset import get_tts_datasets
+from utils.dataset import get_tts_datasets, filter_max_len
 from utils.display import *
 from utils.dsp import np_now
 from utils.duration_extraction import extract_durations_per_count, extract_durations_with_dijkstra
-from utils.files import pickle_binary
+from utils.files import pickle_binary, unpickle_binary
 from utils.metrics import attention_score
 from utils.paths import Paths
 from utils.text import phonemes
 
 
+def normalize_pitch(phoneme_pitches):
+    nonzeros = np.concatenate([v[np.where(v != 0.0)[0]]
+                               for item_id, v in phoneme_pitches])
+    mean, std = np.mean(nonzeros), np.std(nonzeros)
+    for item_id, v in phoneme_pitches:
+        zero_idxs = np.where(v == 0.0)[0]
+        v -= mean
+        v /= std
+        v[zero_idxs] = 0.0
+    return mean, std
+
+
+# adapted from https://github.com/NVIDIA/DeepLearningExamples/blob/
+# 0b27e359a5869cd23294c1707c92f989c0bf201e/PyTorch/SpeechSynthesis/FastPitch/extract_mels.py
+def extract_pitch(save_path: Path) -> Tuple[float, float]:
+    train_data = unpickle_binary('data/train_dataset.pkl')
+    val_data = unpickle_binary('data/val_dataset.pkl')
+    all_data = filter_max_len(train_data + val_data)
+    phoneme_pitches = []
+    for prog_idx, (item_id, mel_len) in enumerate(all_data, 1):
+        dur = np.load(paths.alg / f'{item_id}.npy')
+        assert np.sum(dur) == mel_len
+        pitch = np.load(paths.raw_pitch / f'{item_id}.npy')
+        durs_cum = np.cumsum(np.pad(dur, (1, 0)))
+        pitch_char = np.zeros((dur.shape[0],), dtype=np.float)
+        for idx, a, b in zip(range(mel_len), durs_cum[:-1], durs_cum[1:]):
+            values = pitch[a:b][np.where(pitch[a:b] != 0.0)[0]]
+            values = values[np.where(values < hp.pitch_max_freq)[0]]
+            pitch_char[idx] = np.mean(values) if len(values) > 0 else 0.0
```

Comment:
Hacky? There is a scipy function for this.
```
def _pitch_per_char(self, pitch, durations, mel_len):
    space = np.linspace(0,np.sum(durations), mel_len)
    bin_edges = np.cumsum(np.concatenate([[0], durations]))
    bstat = binned_statistic(space, pitch, bins=bin_edges, statistic='mean')
    return bstat.statistic
```
Naming dur_start and dur_end would be clearer.